### ISSUE TITLE
Crash in failed lax_numpy promotion of int to float

### ISSUE THREAD

**Issue by NeilGirdhar** [google/jax#7367]
Given that `divisor` is a
```
Traced<ShapedArray(int32[], weak_type=True)>with<JVPTrace(level=6/1)>
  with primal = Traced<ShapedArray(int32[], weak_type=True)>with<DynamicJaxprTrace(level=2/1)>
       tangent = Traced<(b'',):JaxprTrace(level=5/1)>
````
Calling `jnp.where(where, divisor, 1.0)` gives:
```
  File "/home/neil/src/tjax/tjax/_src/tools.py", line 62, in divide_where
    divisor = jnp.where(where, divisor, 1.0)
  File "/home/neil/src/jax/jax/_src/numpy/lax_numpy.py", line 1757, in where
    return _where(condition, x, y)
  File "/home/neil/src/jax/jax/_src/numpy/lax_numpy.py", line 1735, in _where
    x, y = _promote_dtypes(x, y)
  File "/home/neil/src/jax/jax/_src/numpy/lax_numpy.py", line 278, in _promote_dtypes
    return [lax._convert_element_type(x, to_dtype, weak_type) for x in args]
  File "/home/neil/src/jax/jax/_src/numpy/lax_numpy.py", line 278, in <listcomp>
    return [lax._convert_element_type(x, to_dtype, weak_type) for x in args]
AssertionError: (array((b'',), dtype=[('float0', 'V')]),)
```

**Comment by mattjj** [comment]
Thanks for raising this! Could you share a repro? The tangent value there is of a placeholder dtype that seems to be leaking where it shouldn't be.

**Comment by NeilGirdhar** [comment]
@mattjj I've been trying to build one.  Can I just give you access to my repo?  You can probably build a MWE much faster than I can.  I wish I knew how to build a `JVPTrace`.  If that's okay, then this should work:

I'm using Python 3.9, so you may have to create a virtual environment.  You can then run it by typing
```sh
pip install poetry
git clone git@github.com:NeilGirdhar/cmm.git
cd cmm
git switch lax_convert_problem
poetry install -E gui
cmm geometry gui
```

**Comment by NeilGirdhar** [comment]
By the way, this may be related to my fix in https://github.com/google/jax/pull/7366, which I needed to get to this point.

**Comment by NeilGirdhar** [comment]
Also, not sure if this is intentional, but `jnp.zeros((), dtype=jax.float0)` doesn't work, but `np.zeros((), dtype=jax.float0)`

**Comment by NeilGirdhar** [comment]
Fixed by @gnecula in https://github.com/google/jax/pull/7378.  Sorry for the second issue.  I really thought this was a separate issue, and was a bit too eager to unblock myself.
